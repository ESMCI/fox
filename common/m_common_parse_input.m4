undefine(`index')dnl
undefine(`len')dnl
undefine(`format')dnl
dnl
include(`foreach.m4')dnl
dnl
include(`common.m4')dnl
dnl
define(`TOHW_defaultargs', `dnl
    character(len=*), intent(in) :: s
    integer, intent(out), optional :: num
    integer, intent(out), optional :: iostat
')dnl
dnl
define(`TOHW_defaultdecls', `dnl
    integer :: i, j, ij, k, s_i, err, ios
    real :: r, c
')dnl
dnl
define(`TOHW_check_errors', `dnl
    if (present(num)) num = ij
    if (ij<size(array)) then
      if (err==0) err = -1
    else
      if (verify(s(s_i:), XML_WHITESPACE)/=0) err = 1
    endif
')dnl
dnl
define(`TOHW_output_errors', `dnl
    if (present(iostat)) then
      iostat = err
    else
      select case (err)
      case(-1)
        write(0, *) "Error in m4f_thisfunc"
        write(0, *) "Too few elements found"
        stop
      case(1)
        write(0, *) "Error in m4f_thisfunc"
        write(0, *) "Too many elements found"
        stop
      case(2)
        write(0, *) "Error in m4f_thisfunc"
        write(0, *) "Malformed input"
        stop
      end select
    end if
')dnl
dnl
define(`TOHW_parse_strings', `dnl
      if (present(separator)) then
        k = index(s(s_i:), separator)
      else
        k = verify(s(s_i:), XML_WHITESPACE)
        if (k==0) exit
        s_i = s_i + k - 1
        k = scan(s(s_i:), XML_WHITESPACE)
      endif
      if (k==0) then
        k = len(s)
      else
        k = s_i + k - 2
      endif
      $1 = s(s_i:k)
      ij = ij + 1
      s_i = k + 2
      if (ij<size(array).and.s_i>len(s)) exit loop
')dnl
dnl
define(`TOHW_parse_logical', `dnl
      k = verify(s(s_i:), XML_WHITESPACE)
      if (k==0) exit
      s_i = s_i + k - 1
      k = scan(s(s_i:), XML_WHITESPACE)
      if (k==0) then
        k = len(s)
      else
        k = s_i + k - 2
      endif
      if (s(s_i:k)=="true".or.s(s_i:k)=="1") then
        $1 = .true.
      elseif (s(s_i:k)=="false".or.s(s_i:k)=="0") then
        $1 = .false.
      else
        err = 2
        exit
      endif
      ij = ij + 1
      s_i = k + 2
      if (ij<size(array).and.s_i>len(s)) exit
')dnl
dnl
define(`TOHW_parse_numbers', `dnl
      k = verify(s(s_i:), XML_WHITESPACE)
      if (k==0) exit
      s_i = s_i + k - 1
      k = scan(s(s_i:), XML_WHITESPACE)
      if (k==0) then
        k = len(s)
      else
        k = s_i + k - 2
      endif
      read(s(s_i:k), *, iostat=ios) $1
      if (ios/=0) then
        err = 2
        exit
      endif
      ij = ij + 1
      s_i = k + 2
      if (ij<size(array).and.s_i>len(s)) exit
')dnl
dnl
define(`TOHW_parse_complex', `dnl
      k = index(s(s_i:), "(")
      if (k==0) exit
      s_i = s_i + k
      k = index(s(s_i:), ")+i(")
      if (k==0) then
        err = 2
        exit
      else
        k = s_i + k - 2
      endif
      ! FIXME should write our own here so it is not recursive I/O
      ! Also not 100% sure that Fortran standard fp read is same as CMLComp
      read(s(s_i:k), *, iostat=ios) r
      if (ios/=0) then
        err = 2
        exit
      endif
      s_i = k + 5
      k = index(s(s_i:), ")")
      if (k==0) then
        err = 2
        exit
      else
        k = s_i + k - 2
      endif
      ! FIXME should write our own here so it is not recursive I/O
      ! Also not 100% sure that Fortran standard fp read is same as CMLComp
      read(s(s_i:k), *, iostat=ios) c
      if (ios/=0) then
        err = 2
        exit
      endif
      $1 = cmplx(r, c)
      ij = ij + 1
      s_i = k + 2
      if (ij<size(array).and.s_i>len(s)) exit
')dnl
dnl
module m_common_parse_input

  use m_common_charset, only : XML_WHITESPACE
  use m_common_realtypes, only: sp, dp

  implicit none
  private

  interface stringtodata
    module procedure arraytostring
    module procedure arraytological
    module procedure arraytointeger
    module procedure arraytorealsp
    module procedure arraytorealdp
    module procedure arraytocomplexsp
    module procedure arraytocomplexdp
    module procedure matrixtostring
    module procedure matrixtological
    module procedure matrixtointeger
    module procedure matrixtorealsp
    module procedure matrixtorealdp
    module procedure matrixtocomplexsp
    module procedure matrixtocomplexdp
  end interface

  public stringtodata

contains

define(`m4f_thisfunc', `arraytostring')dnl
  subroutine m4f_thisfunc`'(s, array, separator, num, iostat)
    character(len=*) :: array(:)
    character, intent(in), optional :: separator
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = ""
    ij = 0
    loop: do i = 1, size(array)
TOHW_parse_strings(`array(i)')
    end do loop

    if (present(num)) num = ij
    if (ij<size(array)) then
      err = -1
    else
      if (present(separator)) then
        if (len(s)-s_i>=0) &
          err = 1
      else
        if (verify(s(s_i:), XML_WHITESPACE)/=0) &
          err = 1
      endif
    endif

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `matrixtostring')dnl
  subroutine m4f_thisfunc`'(s, array, separator, num, iostat)
    character(len=*) :: array(:,:)
    character, intent(in), optional :: separator
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = ""
    ij = 0
    loop: do j = 1, size(array, 2)
    do i = 1, size(array, 1)
TOHW_parse_strings(`array(i, j)')`'dnl
    end do
    end do loop

    if (present(num)) num = ij
    if (ij<size(array)) then
      err = -1
    else
      if (present(separator)) then
        if (len(s)-s_i>=0) &
          err = 1
      else
        if (verify(s(s_i:), XML_WHITESPACE)/=0) &
          err = 1
      endif
    endif

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `arraytological')dnl
  subroutine m4f_thisfunc`'(s, array, num, iostat)
    logical, intent(out) :: array(:)
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = .false.
    ij  = 0
    loop: do i = 1, size(array)
TOHW_parse_logical(`array(i)')
    end do loop

TOHW_check_errors

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `matrixtological')dnl
  subroutine m4f_thisfunc`'(s, array, num, iostat)
    logical, intent(out) :: array(:,:)
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = .false.
    ij = 0
    loop: do j = 1, size(array, 2)
    do i = 1, size(array, 1)
TOHW_parse_logical(`array(i, j)')
    end do
    end do loop

TOHW_check_errors

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `arraytointeger')dnl
  subroutine m4f_thisfunc`'(s, array, num, iostat)
    integer, intent(out) :: array(:)
TOHW_defaultargs

TOHW_defaultdecls
    s_i = 1
    err = 0
    array = 0
    ij  = 0
    loop: do i = 1, size(array)
TOHW_parse_numbers(`array(i)')
    end do loop

TOHW_check_errors

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `matrixtointeger')dnl
  subroutine m4f_thisfunc`'(s, array, num, iostat)
    integer, intent(out) :: array(:, :)
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = 0
    ij = 0
    loop: do j = 1, size(array, 2)
    do i = 1, size(array, 1)
TOHW_parse_numbers(`array(i, j)')
    end do
    end do loop

TOHW_check_errors

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `arraytorealsp')dnl
  subroutine m4f_thisfunc`'(s, array, num, iostat)
    real(sp), intent(out) :: array(:)
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = 0
    ij  = 0
    loop: do i = 1, size(array)
TOHW_parse_numbers(`array(i)')
    end do loop

TOHW_check_errors

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `matrixtorealsp')dnl
  subroutine m4f_thisfunc`'(s, array, num, iostat)
    real(sp), intent(out) :: array(:,:)
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = 0
    ij = 0
    loop: do j = 1, size(array, 2)
    do i = 1, size(array, 1)
TOHW_parse_numbers(`array(i, j)')
    end do
    end do loop

TOHW_check_errors

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `arraytorealdp')dnl
  subroutine m4f_thisfunc`'(s, array, num, iostat)
    real(dp), intent(out) :: array(:)
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = 0
    ij  = 0
    loop: do i = 1, size(array)
TOHW_parse_numbers(`array(i)')
    end do loop

TOHW_check_errors

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `matrixtorealdp')dnl
  subroutine m4f_thisfunc`'(s, array, num, iostat)
    real(dp), intent(out) :: array(:,:)
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = cmplx(0,0)
    ij = 0
    loop: do j = 1, size(array, 2)
    do i = 1, size(array, 1)
TOHW_parse_numbers(`array(i, j)')
    end do
    end do loop

TOHW_check_errors

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `arraytocomplexsp')dnl
  subroutine m4f_thisfunc`'(s, array, num, iostat)
    complex(sp), intent(out) :: array(:)
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = cmplx(0,0)
    ij  = 0
    loop: do i = 1, size(array)
TOHW_parse_complex(`array(i)')
    end do loop

TOHW_check_errors

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `matrixtocomplexsp')dnl
  subroutine m4f_thisfunc`'(s, array, num, iostat)
    complex(sp), intent(out) :: array(:,:)
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = cmplx(0,0)
    ij = 0
    loop: do j = 1, size(array, 2)
    do i = 1, size(array, 1)
TOHW_parse_complex(`array(i, j)')
    end do
    end do loop

TOHW_check_errors

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `arraytocomplexdp')dnl
  subroutine m4f_thisfunc`'(s, array, num, iostat)
    complex(dp), intent(out) :: array(:)
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = cmplx(0)
    ij  = 0
    loop: do i = 1, size(array)
TOHW_parse_complex(`array(i)')
    end do loop

TOHW_check_errors

TOHW_output_errors

  end subroutine m4f_thisfunc

define(`m4f_thisfunc', `matrixtocomplexdp')dnl
  subroutine m4f_thisfunc`'(s, array, num, iostat)
    complex(dp), intent(out) :: array(:,:)
TOHW_defaultargs

TOHW_defaultdecls

    s_i = 1
    err = 0
    array = cmplx(0,0)
    ij = 0
    loop: do j = 1, size(array, 2)
    do i = 1, size(array, 1)
TOHW_parse_complex(`array(i, j)')
    end do
    end do loop

TOHW_check_errors

TOHW_output_errors

  end subroutine m4f_thisfunc
  
end module m_common_parse_input

