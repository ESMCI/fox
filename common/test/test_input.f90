program short_test

  use m_common_parse_input, only : stringtodata

  
  print*,"test.1.1.1"
  call stringdataarray("a b c d e", (/"a", "b", "c", "d", "e"/), 5, 0)
  print*,"test.1.1.2"
  call stringdataarray("a b c d", (/"a", "b", "c", "d", " "/), 4, -1)
  print*,"test.1.1.3"
  call stringdataarray("a b c d e f", (/"a", "b", "c", "d", "e"/), 5, 1)

  print*,"test.1.2.1"
  call stringdataarray("a, b, c, d, e", (/"a ", " b", " c", " d", " e"/), &
    5, 0, ",")
  print*,"test.1.2.2"
  call stringdataarray("a, b, c, d", (/"a ", " b", " c", " d", "  "/), &
    4, -1, ",")
  print*,"test.1.2.3"
  call stringdataarray("a, b, c, d, e, f", (/"a ", " b", " c", " d", " e"/), &
    5, 1, ",")
  print*,"test.1.2.4"
  call stringdataarray("a, b, c, d, e, ", (/"a ", " b", " c", " d", " e"/), &
    5, 1, ",")
  print*,"test.1.2.5"
  call stringdataarray("a, b, c, d, e,", (/"a ", " b", " c", " d", " e"/), &
    5, 0, ",")

  print*,"test.1.3.1"
  call stringdatamatrix("a b c d e f", &
    reshape((/"a", "b", "c", "d", "e", "f"/), (/2,3/)), &
    6, 0)
  print*,"test.1.3.2"
  call stringdatamatrix("a b c d", &
    reshape((/"a", "b", "c", "d", " ", " "/), (/2,3/)), &
    4, -1)
  print*,"test.1.3.3"
  call stringdatamatrix("a b c d e f g", &
    reshape((/"a", "b", "c", "d", "e", "f"/), (/2,3/)), &
    6, 1)

  print*,"test.1.4.1"
  call stringdatamatrix("a, b, c, d, e, f", &
    reshape((/"a ", " b", " c", " d", " e", " f"/), (/2,3/)), &
    6, 0, ",")
  print*,"test.1.4.2"
  call stringdatamatrix("a, b, c, d", &
    reshape((/"a ", " b", " c", " d", "  ", "  "/), (/2,3/)), &
    4, -1, ",")
  print*,"test.1.4.3"
  call stringdatamatrix("a, b, c, d, e, f, g", &
    reshape((/"a ", " b", " c", " d", " e", " f"/), (/2,3/)), &
    6, 1, ",")
  print*,"test.1.4.4"
  call stringdatamatrix("a, b, c, d, e, f, ", &
    reshape((/"a ", " b", " c", " d", " e", " f"/), (/2,3/)), &
    6, 1, ",")
  print*,"test.1.4.5"
  call stringdatamatrix("a, b, c, d, e, f,", &
    reshape((/"a ", " b", " c", " d", " e", " f"/), (/2,3/)), &
    6, 0, ",")

  print*, "test.2.1.1"
  call logicaldataarray("true false false false true", &
    (/.true., .false., .false., .false., .true./), 5, 0)
  print*, "test.2.1.2"
  call logicaldataarray("true false false false", &
    (/.true., .false., .false., .false., .false./), 4, -1)
  print*, "test.2.1.3"
  call logicaldataarray("true false false false true true", &
    (/.true., .false., .false., .false., .true./), 5, 1)

  print*, "test.2.2.1"
  call logicaldatamatrix("true false false false true true", &
    reshape((/.true., .false., .false., .false., .true., .true./), (/2,3/)), &
    6, 0)
  print*, "test.2.2.2"
  call logicaldatamatrix("true false false false", &
    reshape((/.true., .false., .false., .false., .false., .false./), (/2,3/)), &
    4, -1)
  print*, "test.2.2.3"
  call logicaldatamatrix("true false false false true true false", &
    reshape((/.true., .false., .false., .false., .true., .true./), (/2,3/)), &
    6, 1)

  print*, "test.3.1.1"
  call integerdataarray("1 2 3 4 5", &
    (/1, 2, 3, 4, 5/), 5, 0)
  print*, "test.3.1.2"
  call integerdataarray("1 2 3 4", &
    (/1, 2, 3, 4, 0/), 4, -1)
  print*, "test.3.1.3"
  call integerdataarray("1 2 3 4 5 6", &
    (/1, 2, 3, 4, 5/), 5, 1)

! I actually cannot be bothered thinking up test cases
! for real numbers. So shoot me.

  contains

    subroutine stringdataarray(string, array, num, iostat, sep)
      character(len=*), intent(in) :: string
      character(len=*), dimension(:), intent(in) :: array
      integer, intent(in) :: num, iostat
      character, intent(in), optional :: sep

      character(len=len(array)) :: temp(size(array))
      integer :: n, i

      call stringtodata(string, temp, separator=sep, num=n, iostat=i)

      if (any(temp/=array)) &
        print*, "Different array"
      if (i/=iostat) &
        print*, "Wrong iostat"
      if (n/=num) &
        print*, "Wrong num"
    end subroutine stringdataarray

    subroutine stringdatamatrix(string, array, num, iostat, sep)
      character(len=*), intent(in) :: string
      character(len=*), dimension(:, :), intent(in) :: array
      integer, intent(in) :: num, iostat
      character, intent(in), optional :: sep

      character(len=len(array)) :: temp(size(array,1),size(array,2))
      integer :: n, i

      call stringtodata(string, temp, separator=sep, num=n, iostat=i)

      if (any(temp/=array)) &
        print*, "Different array"
      if (i/=iostat) &
        print*, "Wrong iostat"
      if (n/=num) &
        print*, "Wrong num"
    end subroutine stringdatamatrix

    subroutine logicaldataarray(string, array, num, iostat)
      character(len=*), intent(in) :: string
      logical, dimension(:), intent(in) :: array
      integer, intent(in) :: num, iostat

      logical :: temp(size(array))
      integer :: n, i

      call stringtodata(string, temp, n, i)

      if (any(temp.neqv.array)) &
        print*, "Different array"
      if (i/=iostat) &
        print*, "Wrong iostat"
      if (n/=num) &
        print*, "Wrong num"
    end subroutine logicaldataarray

    subroutine logicaldatamatrix(string, array, num, iostat)
      character(len=*), intent(in) :: string
      logical, dimension(:, :), intent(in) :: array
      integer, intent(in) :: num, iostat

      logical :: temp(size(array,1),size(array,2))
      integer :: n, i

      call stringtodata(string, temp, n, i)

      if (any(temp.neqv.array)) &
        print*, "Different array"
      if (i/=iostat) &
        print*, "Wrong iostat"
      if (n/=num) &
        print*, "Wrong num"
    end subroutine logicaldatamatrix

    subroutine integerdataarray(string, array, num, iostat)
      character(len=*), intent(in) :: string
      integer, dimension(:), intent(in) :: array
      integer, intent(in) :: num, iostat

      integer :: temp(size(array))
      integer :: n, i

      call stringtodata(string, temp, n, i)

      if (any(temp/=array)) &
        print*, "Different array"
      if (i/=iostat) &
        print*, "Wrong iostat"
      if (n/=num) &
        print*, "Wrong num"
    end subroutine integerdataarray

    subroutine integerdatamatrix(string, array, num, iostat)
      character(len=*), intent(in) :: string
      integer, dimension(:, :), intent(in) :: array
      integer, intent(in) :: num, iostat

      integer :: temp(size(array,1),size(array,2))
      integer :: n, i

      call stringtodata(string, temp, n, i)

      if (any(temp/=array)) &
        print*, "Different array"
      if (i/=iostat) &
        print*, "Wrong iostat"
      if (n/=num) &
        print*, "Wrong num"
    end subroutine integerdatamatrix

    subroutine realspdataarray(string, array, num, iostat)
      character(len=*), intent(in) :: string
      real, dimension(:), intent(in) :: array
      integer, intent(in) :: num, iostat

      real :: temp(size(array))
      integer :: n, i

      call stringtodata(string, temp, n, i)

      if (any(temp/=array)) &
        print*, "Different array"
      if (i/=iostat) &
        print*, "Wrong iostat"
      if (n/=num) &
        print*, "Wrong num"
    end subroutine realspdataarray

    subroutine realspdatamatrix(string, array, num, iostat)
      character(len=*), intent(in) :: string
      real, dimension(:, :), intent(in) :: array
      integer, intent(in) :: num, iostat

      real :: temp(size(array,1),size(array,2))
      integer :: n, i

      call stringtodata(string, temp, n, i)

      if (any(temp/=array)) &
        print*, "Different array"
      if (i/=iostat) &
        print*, "Wrong iostat"
      if (n/=num) &
        print*, "Wrong num"
    end subroutine realspdatamatrix

end program short_test
