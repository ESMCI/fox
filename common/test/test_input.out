 test.1.1.1
 test.1.1.2
 test.1.1.3
 test.1.2.1
 test.1.2.2
 test.1.2.3
 test.1.2.4
 test.1.2.5
 test.1.3.1
 test.1.3.2
 test.1.3.3
 test.1.4.1
 test.1.4.2
 test.1.4.3
 test.1.4.4
 test.1.4.5
 test.2.1.1
 test.2.1.2
 test.2.1.3
 test.2.2.1
 test.2.2.2
 test.2.2.3
 test.3.1.1
 test.3.1.2
 test.3.1.3
